mvc/models/model-binding.rst

```diff
@@ -1,8 +1,92 @@
+
+How model binding works
+-----------------------
+When ASP.NET MVC receives an incoming HTTP request, it routes it to a specific action method of a controller. It determines which action method to run based on what is in the route data, then it binds values from the HTTP request to that action method's parameters. For example, consider the following URL:
+
+`http://contoso.com/movies/edit/2`
+
+Since the  routing pattern looks like this, ``{controller=Home}/{action=Index}/{id?}``, the route ``movies/edit/2`` maps to the ``Movies`` controller, and its ``Edit`` action method. It also accepts an optional parameter called ``id``. This is due to MVC's use of conventions. So the code for the action method should look something like this: 
+
+.. literalinclude:: model-binding/sample/src/MVCMovie/Controllers/MoviesController.cs
+   :language: c#
+   :lines: 61-76  
+   
+.. note:: The strings in the URL route are not case sensitive.
+
+MVC checks for model binding attributes applied to the action method's arguments. If it finds a model binding attribute it will use the data source specified by the attribute. Otherwise, for each property in the model it queries the value provider for a value with a matching name. In the above example the only matching item is an integer named `id`, so then the value provider associates its values with action method parameters. Below is a list of the data sources in the order that model binding looks through them:
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method.
+#. ``Route values``: Contains information about each parameter of the route. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked the form value provider for a key named `id` and didn't find anything, it moved on to the route values looking for a key with a name of `id`. In our example, it's a match. Binding happens, and the value of the id parameter becomes 2. So far the example use primitive types. If the action method's parameter were a complex type such as the ``Movie`` type, which contains both primitive and complex types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types. While doing so, it compares primitive members' names against the names of keys in data sources and recurses through the properties that are complex types. When a value provider delivers the first item with a matching name from the data source, it stops looking for more items with that name. If binding can't occur, the error is recorded in the ``ModelState.Errors`` collection, and model binding attempts to bind to the next parameter or property. You can query for model state errors by checking the ``ModelState.IsValid`` method. 
```

Comment:
I think the stuff about error handling and model state could probably also go in a separate paragraph.